pkg/daemon/ceph/client/crush_rule.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"fmt"
+	"strconv"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+)
+
+const (
+	crushReplicatedType        = 1
+	ruleMinSizeDefault         = 1
+	ruleMaxSizeDefault         = 10
+	stretchedCRUSHRuleTemplate = `
+rule %s {
+        id %s
+        type replicated
+        min_size %s
+        max_size %s
+        step take %s
+        step choose firstn 0 type %s
+        step chooseleaf firstn 2 type host
+        step emit
+}
+`
+)
+
+var (
+	stepEmit = &stepSpec{Operation: "emit"}
+)
+
+func buildStretchClusterPlainCrushRule(crushMap CrushMap, ruleName string, pool cephv1.PoolSpec) string {
+	return fmt.Sprintf(
+		stretchedCRUSHRuleTemplate,
+		ruleName,
+		strconv.Itoa(crushMap.Rules[len(crushMap.Rules)-1].ID+1),
+		strconv.Itoa(ruleMinSizeDefault),
+		strconv.Itoa(ruleMaxSizeDefault),
+		pool.CrushRoot,
+		pool.FailureDomain,
+	)
+}
+
+func buildStretchClusterCrushRule(crushMap CrushMap, ruleName string, pool cephv1.PoolSpec) *ruleSpec {
+	/*
+		The complete CRUSH rule looks like this:
+
+		   rule two_rep_per_dc {
+		           id 1
+		           type replicated
+		           min_size 1
+		           max_size 10
+		           step take root
+		           step choose firstn 0 type datacenter
+		           step chooseleaf firstn 2 type host
+		           step emit
+		   }
+
+	*/
+
+	ruleID := crushMap.Rules[len(crushMap.Rules)-1].ID + 1
+	return &ruleSpec{
+		ID:      ruleID,
+		Name:    ruleName,
+		Ruleset: ruleID,
+		Type:    crushReplicatedType,
+		MinSize: ruleMinSizeDefault,
+		MaxSize: ruleMaxSizeDefault,
+		Steps:   buildStretchClusterCrushSteps(pool),
+	}
+}
+
+func buildStretchClusterCrushSteps(pool cephv1.PoolSpec) []stepSpec {
+	// Create CRUSH rule steps
+	steps := []stepSpec{}
+
+	// Create the default step, which is essentially the entrypoint, the "root" of all requests
+	stepTakeDefault := &stepSpec{
+		Operation: "take",
+		Item:      -1,
+		ItemName:  pool.CrushRoot,
+	}
+	steps = append(steps, *stepTakeDefault)
+
+	// Steps two
+	stepTakeDatacenter := &stepSpec{
+		Operation: "chooseleaf_firstn",
+		Number:    0,
+		Type:      pool.FailureDomain,
+	}
+	steps = append(steps, *stepTakeDatacenter)
+
+	// Step three
+	stepTakeHost := &stepSpec{
+		Operation: "chooseleaf_firstn",
+		Number:    pool.Replicated.ReplicasPerFailureDomain,
+		Type:      "host",
```

Comment:
Instead of always using host, I wonder if we should allow for another failure domain from another pool property like `subFailureDomain` that is only used when the `replicasPerFailureDomain` is set. But we can certainly delay that feature until there is actually a need...